File: skills/kb-visualize/scripts/graph.py
#!/usr/bin/env python3
"""Extract the graph model of an OKF bundle as JSON — the deterministic half of
kb-visualize. The agent renders this model into a view (HTML artifact or native UI).

Usage:  python3 graph.py <bundle-dir>   # prints graph JSON to stdout

Output:
{
  "bundle": "<dir>",
  "nodes": [ { "id", "path", "type", "title", "description", "tags",
               "resource", "status", "body", "links": [ids], "cited_by": [ids] } ],
  "types": [ "type", ... ],          # distinct types, for coloring/filter
  "edges": [ { "source": id, "target": id } ]
}

Node id = concept id = path within the bundle minus `.md`. Reserved index.md/log.md
are excluded (they are not concepts). Links are resolved to concept ids; links whose
target does not exist in the bundle are dropped from edges (SPEC §5.3 tolerates them).
"""
import os, re, sys, json, posixpath

RESERVED = {"index.md", "log.md"}
FM_RE = re.compile(r"^---\n(.*?)\n---\n?", re.S)
LINK_RE = re.compile(r"\[[^\]]*\]\(([^)#\s]+\.md)(?:#[^)]*)?\)")


def parse_frontmatter(fm):
    """Minimal YAML: scalars and simple `[a, b]` / `- item` lists. No deps."""
    data, key = {}, None
    for line in fm.splitlines():
        if re.match(r"^\s+-\s+", line) and key:
            data.setdefault(key, [])
            if isinstance(data[key], list):
                data[key].append(line.strip()[2:].strip().strip('"\''))
            continue
        m = re.match(r"^([A-Za-z0-9_]+):\s*(.*)$", line)
        if not m:
            continue
        key, val = m.group(1), m.group(2).strip()
        if val == "":
            data[key] = []          # a list follows on subsequent `-` lines
        elif val.startswith("[") and val.endswith("]"):
            data[key] = [x.strip().strip('"\'') for x in val[1:-1].split(",") if x.strip()]
        else:
            data[key] = val.strip('"\'')
    return data


def concept_id(rel):
    return rel[:-3] if rel.endswith(".md") else rel


def resolve(src_rel, target):
    """Resolve a markdown link target (relative or bundle-absolute) to a concept id."""
    if target.startswith("/"):
        resolved = target.lstrip("/")
    else:
        resolved = posixpath.normpath(posixpath.join(posixpath.dirname(src_rel), target))
    return concept_id(resolved)


def build(bundle):
    md = []
    for dp, _, fs in os.walk(bundle):
        for f in fs:
            if f.endswith(".md"):
                md.append(os.path.relpath(os.path.join(dp, f), bundle))

    nodes, ids = {}, set()
    for rel in md:
        if os.path.basename(rel) in RESERVED:
            continue
        ids.add(concept_id(rel))

    for rel in sorted(md):
        if os.path.basename(rel) in RESERVED:
            continue
        text = open(os.path.join(bundle, rel), encoding="utf-8").read()
        m = FM_RE.match(text)
        fm = parse_frontmatter(m.group(1)) if m else {}
        body = text[m.end():] if m else text
        cid = concept_id(rel)
        links = []
        for lm in LINK_RE.finditer(body):
            tgt = lm.group(1)
            if "://" in tgt:
                continue
            rid = resolve(rel, tgt)
            if rid in ids and rid != cid and rid not in links:
                links.append(rid)
        nodes[cid] = {
            "id": cid, "path": rel,
            "type": fm.get("type", ""),
            "title": fm.get("title", cid.rsplit("/", 1)[-1]),
            "description": fm.get("description", ""),
            "tags": fm.get("tags", []) if isinstance(fm.get("tags", []), list) else [fm.get("tags")],
            "resource": fm.get("resource", ""),
            "status": fm.get("status", "active"),
            "body": body.strip(),
            "links": links, "cited_by": [],
        }

    edges = []
    for n in nodes.values():
        for tgt in n["links"]:
            edges.append({"source": n["id"], "target": tgt})
            nodes[tgt]["cited_by"].append(n["id"])

    types = sorted({n["type"] for n in nodes.values() if n["type"]})
    return {"bundle": bundle, "nodes": list(nodes.values()), "types": types, "edges": edges}


def main(argv):
    bundle = next((a for a in argv[1:] if not a.startswith("--")), ".")
    if not os.path.isdir(bundle):
        print(f"not a directory: {bundle}", file=sys.stderr)
        return 2
    print(json.dumps(build(bundle), indent=2))
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: skills/kb-visualize/scripts/graph.mjs
#!/usr/bin/env node

// packages/kb-tools/src/graph.ts
import { readFileSync, statSync as statSync2 } from "node:fs";
import { dirname, join as join2 } from "node:path";

// packages/kb-tools/src/shared.ts
import { readdirSync, statSync } from "node:fs";
import { join, relative, sep } from "node:path";
var FM_RE = /^---\n([\s\S]*?)\n---\n?/;
var RESERVED = /* @__PURE__ */ new Set(["index.md", "log.md"]);
function collectMarkdown(bundle) {
  const out = [];
  const walk = (dir) => {
    let entries;
    try {
      entries = readdirSync(dir).sort();
    } catch {
      return;
    }
    for (const name of entries) {
      const full = join(dir, name);
      let st;
      try {
        st = statSync(full);
      } catch {
        continue;
      }
      if (st.isDirectory()) {
        walk(full);
      } else if (name.endsWith(".md")) {
        out.push(relative(bundle, full).split(sep).join("/"));
      }
    }
  };
  walk(bundle);
  return out;
}
function conceptId(rel) {
  return rel.endsWith(".md") ? rel.slice(0, -3) : rel;
}
var NON_ASCII = new RegExp("[" + String.fromCharCode(128) + "-" + String.fromCharCode(65535) + "]", "g");
function pythonJson(value) {
  return JSON.stringify(value, null, 2).replace(
    NON_ASCII,
    (c) => "\\u" + c.charCodeAt(0).toString(16).padStart(4, "0")
  );
}
function normalizePosix(p) {
  const isAbs = p.startsWith("/");
  const parts = p.split("/");
  const stack = [];
  for (const part of parts) {
    if (part === "" || part === ".") continue;
    if (part === "..") {
      if (stack.length && stack[stack.length - 1] !== "..") stack.pop();
      else if (!isAbs) stack.push("..");
    } else {
      stack.push(part);
    }
  }
  const joined = stack.join("/");
  if (isAbs) return "/" + joined;
  return joined === "" ? "." : joined;
}

// packages/kb-tools/src/graph.ts
var LINK_RE = /\[[^\]]*\]\(([^)#\s]+\.md)(?:#[^)]*)?\)/g;
function stripQuotes(s) {
  return s.replace(/^["']+/, "").replace(/["']+$/, "");
}
function parseFrontmatter(fm) {
  const data = {};
  let key = null;
  for (const line of fm.split("\n")) {
    if (/^\s+-\s+/.test(line) && key) {
      if (!(key in data)) data[key] = [];
      const cur = data[key];
      if (Array.isArray(cur)) {
        cur.push(stripQuotes(line.trim().slice(2).trim()));
      }
      continue;
    }
    const m = /^([A-Za-z0-9_]+):\s*(.*)$/.exec(line);
    if (!m) continue;
    key = m[1];
    const val = m[2].trim();
    if (val === "") {
      data[key] = [];
    } else if (val.startsWith("[") && val.endsWith("]")) {
      data[key] = val.slice(1, -1).split(",").map((x) => x.trim()).filter((x) => x.length > 0).map(stripQuotes);
    } else {
      data[key] = stripQuotes(val);
    }
  }
  return data;
}
function scalar(data, k, dflt) {
  const v = data[k];
  if (v === void 0) return dflt;
  return Array.isArray(v) ? dflt : v;
}
function resolve(srcRel, target) {
  const resolved = target.startsWith("/") ? target.replace(/^\/+/, "") : normalizePosix(`${dirname(srcRel) === "." ? "" : dirname(srcRel)}/${target}`.replace(/^\//, ""));
  return conceptId(resolved);
}
function extractGraph(bundle) {
  const md = collectMarkdown(bundle);
  const posixBasename = (rel) => rel.split("/").pop() ?? rel;
  const ids = /* @__PURE__ */ new Set();
  for (const rel of md) {
    if (RESERVED.has(posixBasename(rel))) continue;
    ids.add(conceptId(rel));
  }
  const nodes = /* @__PURE__ */ new Map();
  for (const rel of [...md].sort()) {
    if (RESERVED.has(posixBasename(rel))) continue;
    const text = readFileSync(join2(bundle, rel), "utf-8");
    const m = FM_RE.exec(text);
    const fm = m ? parseFrontmatter(m[1]) : {};
    const body = m ? text.slice(m[0].length) : text;
    const cid = conceptId(rel);
    const links = [];
    for (const lm of body.matchAll(LINK_RE)) {
      const tgt = lm[1];
      if (tgt.includes("://")) continue;
      const rid = resolve(rel, tgt);
      if (ids.has(rid) && rid !== cid && !links.includes(rid)) links.push(rid);
    }
    const rawTags = fm["tags"];
    const tags = Array.isArray(rawTags) ? rawTags : rawTags === void 0 ? [] : [rawTags];
    nodes.set(cid, {
      id: cid,
      path: rel,
      type: scalar(fm, "type", ""),
      title: scalar(fm, "title", cid.split("/").pop() ?? cid),
      description: scalar(fm, "description", ""),
      tags,
      resource: scalar(fm, "resource", ""),
      status: scalar(fm, "status", "active"),
      body: body.trim(),
      links,
      cited_by: []
    });
  }
  const edges = [];
  for (const n of nodes.values()) {
    for (const tgt of n.links) {
      edges.push({ source: n.id, target: tgt });
      nodes.get(tgt).cited_by.push(n.id);
    }
  }
  const types = [...new Set([...nodes.values()].map((n) => n.type).filter((t) => t))].sort();
  return { bundle, nodes: [...nodes.values()], types, edges };
}
function runCli(argv) {
  const bundle = argv.find((a) => !a.startsWith("--")) ?? ".";
  let isDir = false;
  try {
    isDir = statSync2(bundle).isDirectory();
  } catch {
    isDir = false;
  }
  if (!isDir) {
    process.stderr.write(`not a directory: ${bundle}
`);
    return 2;
  }
  process.stdout.write(pythonJson(extractGraph(bundle)) + "\n");
  return 0;
}

// packages/kb-tools/src/cli/graph-cli.ts
process.exit(runCli(process.argv.slice(2)));
